File: packages/js-sdk/src/http-client.ts
/**
 * HTTP Client for DeerFlow API
 */

import type { ClientConfig } from "./types";

export class HttpClient {
  private baseUrl: string;
  private authToken?: string;
  private userId?: string;
  private timeout: number;

  constructor(config: ClientConfig) {
    this.baseUrl = config.baseUrl.replace(/\/$/, "");
    this.authToken = config.authToken;
    this.userId = config.userId;
    this.timeout = config.timeout ?? 30000;
  }

  /**
   * Build a request URL with optional query parameters
   */
  private buildUrl(path: string, params?: Record<string, string | number | undefined>): string {
    const url = new URL(`${this.baseUrl}${path}`);
    if (params) {
      for (const [key, value] of Object.entries(params)) {
        if (value !== undefined) {
          url.searchParams.set(key, String(value));
        }
      }
    }
    return url.toString();
  }

  /**
   * Get default headers for requests
   */
  getDefaultHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
    };

    if (this.authToken) {
      headers["Authorization"] = `Bearer ${this.authToken}`;
    }

    return headers;
  }

  /**
   * Get headers with CSRF token for state-changing requests
   */
  private getAuthHeaders(): Record<string, string> {
    const headers = this.getDefaultHeaders();
    // CSRF token handling is done at the browser level via cookies
    // This is for internal API calls
    if (this.authToken) {
      headers["X-Internal-Auth"] = this.authToken;
    }
    return headers;
  }

  /**
   * Handle response errors
   */
  private async handleResponse<T>(response: Response): Promise<T> {
    if (!response.ok) {
      let errorData: unknown;
      try {
        errorData = await response.json();
      } catch {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }

      const message = (errorData as { detail?: string })?.detail || `HTTP ${response.status}`;
      throw new Error(message);
    }

    // Handle empty responses
    const contentType = response.headers.get("content-type");
    if (!contentType || !contentType.includes("application/json")) {
      return {} as T;
    }

    return response.json() as Promise<T>;
  }

  /**
   * Fetch with timeout
   */
  private async fetchWithTimeout(url: string, options: RequestInit = {}): Promise<Response> {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(url, {
        ...options,
        signal: options.signal || controller.signal,
      });
      return response;
    } finally {
      clearTimeout(timeoutId);
    }
  }

  /**
   * GET request
   */
  async get<T>(path: string, params?: Record<string, string | number | undefined>): Promise<T> {
    const url = this.buildUrl(path, params);
    const response = await this.fetchWithTimeout(url, {
      method: "GET",
      headers: this.getDefaultHeaders(),
    });
    return this.handleResponse<T>(response);
  }

  /**
   * POST request
   */
  async post<T>(path: string, body?: unknown, headers?: HeadersInit): Promise<T> {
    const response = await this.fetchWithTimeout(`${this.baseUrl}${path}`, {
      method: "POST",
      headers: this.getAuthHeaders(),
      body: body !== undefined ? JSON.stringify(body) : undefined,
    });
    return this.handleResponse<T>(response);
  }

  /**
   * PUT request
   */
  async put<T>(path: string, body?: unknown): Promise<T> {
    const response = await this.fetchWithTimeout(`${this.baseUrl}${path}`, {
      method: "PUT",
      headers: this.getAuthHeaders(),
      body: body !== undefined ? JSON.stringify(body) : undefined,
    });
    return this.handleResponse<T>(response);
  }

  /**
   * DELETE request
   */
  async delete<T>(path: string): Promise<T> {
    const response = await this.fetchWithTimeout(`${this.baseUrl}${path}`, {
      method: "DELETE",
      headers: this.getAuthHeaders(),
    });
    return this.handleResponse<T>(response);
  }

  /**
   * Upload files (multipart/form-data)
   */
  async upload<T>(path: string, formData: FormData): Promise<T> {
    const response = await this.fetchWithTimeout(`${this.baseUrl}${path}`, {
      method: "POST",
      headers: this.authToken ? { Authorization: `Bearer ${this.authToken}` } : {},
      body: formData,
    });
    return this.handleResponse<T>(response);
  }

  /**
   * Download file as blob
   */
  async download(path: string): Promise<Blob> {
    const response = await this.fetchWithTimeout(`${this.baseUrl}${path}`, {
      method: "GET",
      headers: this.getDefaultHeaders(),
    });

    if (!response.ok) {
      throw new Error(`Download failed: ${response.status}`);
    }

    return response.blob();
  }

  /**
   * Get the base URL
   */
  getBaseUrl(): string {
    return this.baseUrl;
  }

  /**
   * SSE stream request
   */
  async stream(path: string, body?: unknown, onEvent?: (event: string, data: unknown) => void): Promise<void> {
    const url = `${this.baseUrl}${path}`;

    const response = await fetch(url, {
      method: "POST",
      headers: this.getAuthHeaders(),
      body: body !== undefined ? JSON.stringify(body) : undefined,
    });

    if (!response.ok) {
      throw new Error(`Stream request failed: ${response.status}`);
    }

    if (!response.body) {
      throw new Error("Response body is null");
    }

    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() || "";

      for (const line of lines) {
        if (line.startsWith("data: ")) {
          const data = line.slice(6);
          if (onEvent) {
            try {
              const parsed = JSON.parse(data);
              onEvent("message", parsed);
            } catch {
              onEvent("message", data);
            }
          }
        }
      }
    }
  }
}

File: packages/android-sdk/src/com/deerflow/android/sdk/HttpClient.java
package com.deerflow.android.sdk;

import com.deerflow.android.sdk.models.*;
import com.deerflow.android.sdk.utils.JsonUtils;

import okhttp3.*;
import okhttp3.MediaType;
import okhttp3.RequestBody;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.File;
import java.io.IOException;
import java.util.*;
import java.util.concurrent.TimeUnit;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

/**
 * HTTP Client for DeerFlow API using OkHttp
 */
public class HttpClient {

    private final OkHttpClient client;
    private final String baseUrl;
    private final String authToken;
    private final String userId;

    public HttpClient(ClientConfig config) {
        this.baseUrl = config.getBaseUrl().replaceAll("/$", "");
        this.authToken = config.getAuthToken();
        this.userId = config.getUserId();

        this.client = new OkHttpClient.Builder()
                .connectTimeout(config.getTimeout(), TimeUnit.MILLISECONDS)
                .readTimeout(config.getTimeout(), TimeUnit.MILLISECONDS)
                .writeTimeout(config.getTimeout(), TimeUnit.MILLISECONDS)
                .build();
    }

    /**
     * Get default headers for requests
     */
    private Headers getDefaultHeaders() {
        Headers.Builder builder = new Headers.Builder()
                .add("Content-Type", "application/json");
        if (authToken != null) {
            builder.add("Authorization", "Bearer " + authToken);
        }
        return builder.build();
    }

    /**
     * Get auth headers for state-changing requests
     */
    private Headers getAuthHeaders() {
        Headers.Builder builder = new Headers.Builder();
        builder.add("Content-Type", "application/json");
        if (authToken != null) {
            builder.add("Authorization", "Bearer " + authToken);
            builder.add("X-Internal-Auth", authToken);
        }
        return builder.build();
    }

    /**
     * Handle response and throw exception on error
     */
    private <T> T handleResponse(Response response, Class<T> clazz) throws Exception {
        if (!response.isSuccessful()) {
            String errorBody = response.body() != null ? response.body().string() : "";
            String message = errorBody.isEmpty() ? "HTTP " + response.code() : errorBody;

            switch (response.code()) {
                case 401:
                    throw new AuthenticationException(message);
                case 404:
                    throw new NotFoundException(message);
                default:
                    throw new DeerFlowException(message, null, response.code());
            }
        }

        String body = response.body() != null ? response.body().string() : "";
        if (clazz == String.class) {
            return clazz.cast(body);
        }
        return JsonUtils.fromJson(body, clazz);
    }

    /**
     * GET request
     */
    public <T> T get(String path, Map<String, Object> params, Class<T> clazz) throws Exception {
        String url = buildUrl(path, params);
        Request request = new Request.Builder()
                .url(url)
                .get()
                .headers(getDefaultHeaders())
                .build();

        try (Response response = client.newCall(request).execute()) {
            return handleResponse(response, clazz);
        }
    }

    /**
     * POST request
     */
    public <T> T post(String path, Object body, Class<T> clazz) throws Exception {
        String jsonBody = body != null ? JsonUtils.toJson(body) : "{}";
        RequestBody requestBody = RequestBody.create(jsonBody, MediaType.parse("application/json"));

        Request request = new Request.Builder()
                .url(baseUrl + path)
                .post(requestBody)
                .headers(getAuthHeaders())
                .build();

        try (Response response = client.newCall(request).execute()) {
            return handleResponse(response, clazz);
        }
    }

    /**
     * PUT request
     */
    public <T> T put(String path, Object body, Class<T> clazz) throws Exception {
        String jsonBody = body != null ? JsonUtils.toJson(body) : "{}";
        RequestBody requestBody = RequestBody.create(jsonBody, MediaType.parse("application/json"));

        Request request = new Request.Builder()
                .url(baseUrl + path)
                .put(requestBody)
                .headers(getAuthHeaders())
                .build();

        try (Response response = client.newCall(request).execute()) {
            return handleResponse(response, clazz);
        }
    }

    /**
     * DELETE request
     */
    public <T> T delete(String path, Class<T> clazz) throws Exception {
        Request request = new Request.Builder()
                .url(baseUrl + path)
                .delete()
                .headers(getAuthHeaders())
                .build();

        try (Response response = client.newCall(request).execute()) {
            return handleResponse(response, clazz);
        }
    }

    /**
     * Upload file
     */
    public <T> T upload(String path, File file, Class<T> clazz) throws Exception {
        RequestBody fileBody = RequestBody.create(file, MediaType.parse("application/octet-stream"));
        RequestBody requestBody = new MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("files", file.getName(), fileBody)
                .build();

        Request.Builder requestBuilder = new Request.Builder()
                .url(baseUrl + path)
                .post(requestBody);

        if (authToken != null) {
            requestBuilder.addHeader("Authorization", "Bearer " + authToken);
        }

        try (Response response = client.newCall(requestBuilder.build()).execute()) {
            return handleResponse(response, clazz);
        }
    }

    /**
     * Download file as bytes
     */
    public byte[] download(String path) throws Exception {
        Request request = new Request.Builder()
                .url(baseUrl + path)
                .get()
                .headers(getDefaultHeaders())
                .build();

        try (Response response = client.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                throw new DeerFlowException("Download failed: " + response.code(), null, response.code());
            }
            byte[] bytes = response.body() != null ? response.body().bytes() : new byte[0];
            if (bytes.length == 0) {
                throw new IOException("Empty response");
            }
            return bytes;
        }
    }

    /**
     * Stream request with callback
     */
    public void stream(String path, Object body, BiConsumer<String, String> onEvent) throws Exception {
        String jsonBody = body != null ? JsonUtils.toJson(body) : "{}";
        RequestBody requestBody = RequestBody.create(jsonBody, MediaType.parse("application/json"));

        Request request = new Request.Builder()
                .url(baseUrl + path)
                .post(requestBody)
                .headers(getAuthHeaders())
                .build();

        try (Response response = client.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                throw new DeerFlowException("Stream request failed: " + response.code(), null, response.code());
            }

            if (response.body() == null) {
                throw new IOException("Empty response body");
            }

            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(response.body().byteStream()))) {
                String currentEvent = "";
                String line;

                while ((line = reader.readLine()) != null) {
                    if (line.startsWith("event: ")) {
                        currentEvent = line.substring(7).trim();
                    } else if (line.startsWith("data: ")) {
                        String data = line.substring(6);
                        onEvent.accept(currentEvent, data);
                    }
                }
            }
        }
    }

    /**
     * Build URL with query parameters
     */
    private String buildUrl(String path, Map<String, Object> params) {
        StringBuilder urlBuilder = new StringBuilder(baseUrl + path);
        if (params != null && !params.isEmpty()) {
            List<String> pairs = new ArrayList<>();
            for (Map.Entry<String, Object> entry : params.entrySet()) {
                if (entry.getValue() != null) {
                    try {
                        String encoded = java.net.URLEncoder.encode(entry.getValue().toString(), "UTF-8");
                        pairs.add(entry.getKey() + "=" + encoded);
                    } catch (Exception e) {
                        pairs.add(entry.getKey() + "=" + entry.getValue());
                    }
                }
            }
            if (!pairs.isEmpty()) {
                urlBuilder.append("?").append(String.join("&", pairs));
            }
        }
        return urlBuilder.toString();
    }

    /**
     * Get base URL
     */
    public String getBaseUrl() {
        return baseUrl;
    }
}
